src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -206,37 +212,44 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
 object ArrayBuffer extends StrictOptimizedSeqFactory[ArrayBuffer] {
 
   // Avoid reallocation of buffer if length is known.
-  def from[B](coll: collection.IterableOnce[B]): ArrayBuffer[B] =
-    if (coll.knownSize >= 0) {
-      val array = new Array[AnyRef](coll.knownSize)
+  def from[B](coll: collection.IterableOnce[B]): ArrayBuffer[B] = {
+    val k = coll.knownSize
+    if (k >= 0) {
+      val array = new Array[AnyRef](k max minimalLength)
       val it = coll.iterator
-      for (i <- 0 until array.length) array(i) = it.next().asInstanceOf[AnyRef]
-      new ArrayBuffer[B](array, array.length)
+      for (i <- 0 until k) array(i) = it.next().asInstanceOf[AnyRef]
+      new ArrayBuffer[B](array, k)
     }
     else new ArrayBuffer[B] ++= coll
+  }
 
   def newBuilder[A]: Builder[A, ArrayBuffer[A]] =
     new GrowableBuilder[A, ArrayBuffer[A]](empty) {
       override def sizeHint(size: Int): Unit = elems.ensureSize(size)
     }
 
   def empty[A]: ArrayBuffer[A] = new ArrayBuffer[A]()
-}
-
-final class ArrayBufferView[A](val array: Array[AnyRef], val length: Int) extends AbstractIndexedSeqView[A] {
-  @throws[ArrayIndexOutOfBoundsException]
-  def apply(n: Int) = if (n < length) array(n).asInstanceOf[A] else throw new IndexOutOfBoundsException(s"$n is out of bounds (min 0, max ${length - 1})")
-  override protected[this] def className = "ArrayBufferView"
-}
 
-/** An object used internally by collections backed by an extensible Array[AnyRef] */
-object RefArrayUtils {
+  // We chose 8 as a good size for the minimum array
+  private[this] val minimalLength: Int = 8
+
+  private def reduceSize(array: Array[AnyRef], n: Int): Array[AnyRef] = {
+    var newLength: Long = array.length
+    if(newLength == Int.MaxValue) newLength += 1 // ensure that newLength is a power of 2
+    val minLength = minimalLength max n
+    while(newLength/2 >= minLength) newLength /= 2
+    if (newLength != array.length) {
```

Comment:
`newLength != array.length && newLength < Int.MaxValue`

1 was added to newLength earlier, so if in case your `n` is close to `Int.MaxValue`, then this would called `(Int.MaxValue + 1L).toInt`, which is `-2147483648`.